### PR TITLE
Publish JavaDoc on build

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR brings the scripts publishing the generated JavaDoc to [spine.io](https://spine.io/core-java/javadoc/server/) via GitHub Pages.

The JavaDoc is generated locally and added to the `gh-pages` branch, which is automatically served under the specified URL.

Note that only the non-`@Internal` JavaDoc is published.

To publish the docs, run `./gradlew updateGitHubPages`.

**Warning:** it is not recommended to publish the doc from your local environment. The publishing works automatically when merging a branch into `master`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: .travis.yml [hunk 1]
```diff
@@ -10,9 +10,15 @@ env:
...
     # Encrypted `GCS_SECRET` variable.
     - secure: MDtkG8hUJtC67lf3jYwS4nkCScTTWq2kNGvOK5KaRYgvLKAVd6pIxE0OSUR7IJALLu5O4xAWhJH26TU5QcVrmnRMY4MrhLDgjDfcYCxoclPzTRzxlsGvZz2For02kaHxMF7CXzdtr13zr8EZR5nQI6uQoY39R6w01/xD7lWYAz5WEwlCy700AuOcZ5u4I4Me+9vhu2BO1fYCp1Enq4xkk6ARGXqF7MTm4fMuZeMLOA6PlQ94RaM+XMQ8VKqyYamv46O5P95/SwtNKfeFUCe+NIKVOX23VcS1om02KRNhYmVoABco0BNX3oNRqbVcXkubvAAgg5OHM+k5DbRk306WnMxvlhMitNQeJQ0xGuEPIjCREoZw3AXh+iV5xmaQ1EUa8tsJHMpLishsu9S/towRjtM+r/ODiLFWh+caBqKyDyExwrM79XOuv1/UpV9ZB5/eoz5oqyrO9NNJpJVmDkffh3NYXavE0Zb48g+1NLCvUPgIu7i76JCtK7y5BwYQQZqhOlU/QfF8TDekFv31JecSMpg3zwTGcec8a9F19J+G8bliheDSzpQ6l1AaRL6glzBWrGdGHMLRYJDYi8bHz/uYqavb8GMg61+kMTkXxgtSFCs47voJ/7HyMTOLNGJ+DCBb94GkqWcGrjsO3W1lNoKTmgLWXGU8ALMvqYDVTDolvEI=
+    # The email attached to the commits which update the JavaDoc hosted on GitHub Pages.
+    # See `scripts/update-gh-pages.gradle`
+    - FORMAL_GIT_HUB_PAGES_AUTHOR: "dmitry.dashenkov@teamdev.com"
```
Comment (armiol) [inline .travis.yml]: Ideally we should have the actual person who performs the changes.

If we cannot, then we need to have a Google Group email instead. Something like `gh-publisher@...`. Let's discuss.

File: build.gradle [hunk 2]
```diff
@@ -100,6 +100,19 @@ allprojects {
 
 final boolean isTravis = System.env.TRAVIS == 'true'
 
+/**
+ * Defines whether the given {@code project} should expose its JavaDoc to spine.io Web site.
```
Comment (armiol) [inline build.gradle]: I think, it's `Determines`, `Javadoc` and `website`.

Let's have `Javadoc` spelling through this changeset.

File: build.gradle [hunk 3]
```diff
@@ -100,6 +100,19 @@ allprojects {
...
+ *
+ * <p>Currently, the {@code testutil} projects are excluded from publishing, as well as the model
+ * compile time checking modules.
```
Comment (armiol) [inline build.gradle]: `model compile time checking modules` blows minds away.

File: build.gradle [hunk 4]
```diff
@@ -327,6 +340,12 @@ subprojects {
             }
         }
     }
+
+    // Apply the JavaDoc publishing plugin.
+    // This plugin *must* be applied here, not in the module `build.gradle` files.
```
Comment (armiol) [inline build.gradle]: Please have an `//` (empty comment line) above.

File: scripts/update-gh-pages.gradle [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    final def errWriter = new StringWriter()
+    process.consumeProcessOutputStream(outWriter)
+    process.consumeProcessErrorStream(errWriter)
+    final int exitCode = process.waitFor()
+    final def result = [out: outWriter.toString(), error: errWriter.toString(), exitCode: exitCode]
+    return result
+}
+
+/**
+ * Executes the given command and retrieves the command output.
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: I think that's about OS terminal commands. Please be more specific.

File: scripts/update-gh-pages.gradle [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        final String errorMsg = "Command `$command` finished with exit code $result.exitCode:" +
+                " ${System.lineSeparator()}${result.error}"
+        throw new IllegalStateException(errorMsg)
+    }
+}
+
+/**
+ * Executes the given command and retrieves the command output.
+ *
+ * <p>This is a convenience method. Acts as if {@code execute(project.rootDir, command)}
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: `as if` is inapplicable without a verb.

File: scripts/update-gh-pages.gradle [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    execute repoBaseDir, "git", "add", javaDocDirPostfix
+
+    /*
+     * Publish the changes under "UpdateGitHubPages Plugin" Git user name and email stored in
+     * "FORMAL_GIT_HUB_PAGES_AUTHOR" env variable.
+     *
+     * When changing the value of "FORMAL_GIT_HUB_PAGES_AUTHOR", also change the SSH private
+      * (deploy_key_rsa.enc) and public ("GitHub Pages publisher (Travis CI)" on GitHub) keys.
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: An extra space symbol is here.

File: scripts/update-gh-pages.gradle [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        into javaDocDir
+    }
+
+    execute repoBaseDir, "git", "add", javaDocDirPostfix
+
+    /*
+     * Publish the changes under "UpdateGitHubPages Plugin" Git user name and email stored in
+     * "FORMAL_GIT_HUB_PAGES_AUTHOR" env variable.
+     *
+     * When changing the value of "FORMAL_GIT_HUB_PAGES_AUTHOR", also change the SSH private
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: Let's use `<p>` consistently for new paragraphs.

File: build.gradle (hunk 2 above)
Comment (armiol) [inline build.gradle]: I also think, that that's not `spine.io` website, but `SpineEventEngine.github.io`.

File: .travis.yml [hunk 9]
```diff
@@ -10,9 +10,15 @@ env:
...
     # Encrypted `GCS_SECRET` variable.
     - secure: MDtkG8hUJtC67lf3jYwS4nkCScTTWq2kNGvOK5KaRYgvLKAVd6pIxE0OSUR7IJALLu5O4xAWhJH26TU5QcVrmnRMY4MrhLDgjDfcYCxoclPzTRzxlsGvZz2For02kaHxMF7CXzdtr13zr8EZR5nQI6uQoY39R6w01/xD7lWYAz5WEwlCy700AuOcZ5u4I4Me+9vhu2BO1fYCp1Enq4xkk6ARGXqF7MTm4fMuZeMLOA6PlQ94RaM+XMQ8VKqyYamv46O5P95/SwtNKfeFUCe+NIKVOX23VcS1om02KRNhYmVoABco0BNX3oNRqbVcXkubvAAgg5OHM+k5DbRk306WnMxvlhMitNQeJQ0xGuEPIjCREoZw3AXh+iV5xmaQ1EUa8tsJHMpLishsu9S/towRjtM+r/ODiLFWh+caBqKyDyExwrM79XOuv1/UpV9ZB5/eoz5oqyrO9NNJpJVmDkffh3NYXavE0Zb48g+1NLCvUPgIu7i76JCtK7y5BwYQQZqhOlU/QfF8TDekFv31JecSMpg3zwTGcec8a9F19J+G8bliheDSzpQ6l1AaRL6glzBWrGdGHMLRYJDYi8bHz/uYqavb8GMg61+kMTkXxgtSFCs47voJ/7HyMTOLNGJ+DCBb94GkqWcGrjsO3W1lNoKTmgLWXGU8ALMvqYDVTDolvEI=
+    # The email attached to the commits which update the Javadoc hosted on GitHub Pages.
```
Comment (armiol) [inline .travis.yml]: `This email is used as a commit author email when updating the Javadoc hosted on GitHub Pages.`

File: scripts/update-gh-pages.gradle [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+ * The GitHub URL to the {@code core-java} project.
+ *
+ * <p>The host name is {@code github.com-publish}, though resolved to {@code github.com} by SSH on
+ * Travis due to the {@code .travis_ssh_config} configuration. This is used to make it possible
+ * to have multiple SSH private keys on Travis (since there is one already on Travis).
+ */
+final String gitHost = "git@github.com-publish:SpineEventEngine/core-java.git"
+
+/**
+ * The GitHub deploy key used to push the Javadoc changes to {@code gh-pages}.
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: `... to the {@code gh-pages} branch.`

File: scripts/update-gh-pages.gradle (hunk 8 above)
Comment (armiol) [reply scripts/update-gh-pages.gradle]: Please have a resolution for this one written down.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: scripts/update-gh-pages.gradle [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Executes the given terminal command and retrieves the command output.
+ *
+ * <p>{@link Runtime#exec(String[], String[], File) Executes} the given {@code String} array as
+ * a CLI command. If the execution is successful, returns the command output. Thrown
+ * an {@link IllegalStateException} otherwise.
+ *
+ * @param fromDir the directory of the command execution
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: `baseDir` is more common.

File: scripts/update-gh-pages.gradle [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * The GitHub URL to the {@code core-java} project.
+ *
+ * <p>The host name is {@code github.com-publish}, though resolved to {@code github.com} by SSH on
+ * Travis due to the {@code .travis_ssh_config} configuration. This is used to make it possible
+ * to have multiple SSH private keys on Travis (since there is one already on Travis).
+ */
+final String gitHost = "git@github.com-publish:SpineEventEngine/core-java.git"
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: Is this a constant? If so, let's name it accordingly. Same below.

File: scripts/update-gh-pages.gradle [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    final int exitCode = process.waitFor()
+    final def result = [out: outWriter.toString(), error: errWriter.toString(), exitCode: exitCode]
+    return result
+}
+
+/**
+ * Executes the given terminal command and retrieves the command output.
+ *
+ * <p>{@link Runtime#exec(String[], String[], File) Executes} the given {@code String} array as
+ * a CLI command. If the execution is successful, returns the command output. Thrown
```
Comment (armiol) [inline scripts/update-gh-pages.gradle]: `Thrown` seems misleading.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Pull request merged**